go/libraries/doltcore/merge/type_compatibility.go

```diff
@@ -29,142 +30,260 @@ import (
+	// isCompatible returns two booleans – the first boolean response parameter indicates if a type change from
+	// |fromType| to |toType| is compatible and safe to perform automatically. The second boolean response parameter
+	// indicates if the type conversion requires a full table rewrite.
+	isCompatible(fromSqlType, toSqlType sql.Type) (bool, bool)
+}
 
-		// MySQL uses 1 byte to store enum values that have <= 255 values, and 2 bytes for > 255 values
-		// The DOLT storage format *always* uses 2 bytes for all enum values, so table data never needs
-		// to be rewritten in this additive case.
-		return true
+// stringTypeChangeHandler handles type change compatibility between from string types, i.e. VARCHAR, VARBINARY,
+// CHAR, BINARY, TEXT, and BLOB types.
+type stringTypeChangeHandler struct{}
 
-	case types.IsSet(fromSqlType) && types.IsSet(toSqlType):
-		fromSetType := fromSqlType.(sql.SetType)
-		toSetType := toSqlType.(sql.SetType)
-		if fromSetType.NumberOfElements() > toSetType.NumberOfElements() {
+var _ typeChangeHandler = (*stringTypeChangeHandler)(nil)
+
+func (s stringTypeChangeHandler) canHandle(fromSqlType, toSqlType sql.Type) bool {
+	switch fromSqlType.Type() {
+	case query.Type_VARCHAR, query.Type_CHAR, query.Type_TEXT:
+		switch toSqlType.Type() {
+		case query.Type_VARCHAR, query.Type_CHAR, query.Type_TEXT:
+			return true
+		default:
 			return false
 		}
-
-		// TODO: charset/collation changes may require a table or index rewrite; for now, consider them incompatible
-		fromCharSet, fromCollation := fromSetType.CharacterSet(), fromSetType.Collation()
-		toCharSet, toCollation := toSetType.CharacterSet(), toSetType.Collation()
-		if fromCharSet != toCharSet || fromCollation != toCollation {
+	case query.Type_VARBINARY, query.Type_BINARY, query.Type_BLOB:
+		switch toSqlType.Type() {
+		case query.Type_VARBINARY, query.Type_BINARY, query.Type_BLOB:
+			return true
+		default:
 			return false
 		}
+	default:
+		return false
+	}
+}
 
-		// Ensure only new values have been added to the end of the set
-		toSetValues := toSetType.Values()
-		for i, fromSetValue := range fromSetType.Values() {
-			if toSetValues[i] != fromSetValue {
-				return false
-			}
+func (s stringTypeChangeHandler) isCompatible(fromSqlType, toSqlType sql.Type) (bool, bool) {
+	fromStringType := fromSqlType.(types.StringType)
+	toStringType := toSqlType.(types.StringType)
+
+	compatible := toStringType.MaxByteLength() >= fromStringType.MaxByteLength() &&
+		toStringType.Collation() == fromStringType.Collation()
+
+	rewriteRequired := false
+	if compatible {
+		// Because inline string types (e.g. VARCHAR, CHAR) have the same encoding, the only time
+		// a table rewrite is required is when moving between an inline string type and an
+		// out-of-band string type (e.g. TEXT).
+		fromTypeOutOfBand := outOfBandType(fromSqlType)
+		toTypeOutOfBand := outOfBandType(toSqlType)
+		if !fromTypeOutOfBand && toTypeOutOfBand {
```

Comment:
Isn't the same true in the other direction as well (TEXT -> VARCHAR)

Should be `fromTypeOutOfBand != toTypeOutOfBand`?

Reply:
Test for this case?

Reply:
Great catch! I was originally assuming that any TEXT to VARCHAR conversion would error with a schema conflict and not get this far, but TINYTEXT to VARCHAR(300) is totally valid, and does require a table rewrite. Fixed in commit 
133c88b4568152da03dadb678bdcf2310522dcaa and added more tests for this.